### ISSUE TITLE
Inefficiency in generating initial conditions while using optimize_acqf with fixed_features

### ISSUE THREAD

**Issue by saitcakmak** [pytorch/botorch#345]
## Issue description

While using `optimize_acqf` with `fixed_features`, the initial condition generation does not account for the `fixed_features`. The `fixed_features` are later applied during optimization with `gen_candidates_scipy`. I believe this could be improved by using `fixed_features` during initial condition generation to improve the quality of the initial conditions. 

## Code example

Following code can be used in debug mode along with a breakpoint at the `forward` pass of `qKnowledgeGradient` to observe the behavior. The first call of `forward` will not have `X[:, :, 0] = 0.5` fixed (this is the initial condition generation), later calls will have it fixed.

``` python
import torch
from botorch.acquisition import qKnowledgeGradient
from botorch.fit import fit_gpytorch_model
from botorch.models import SingleTaskGP
from botorch.models.transforms import Standardize
from botorch.optim import optimize_acqf
from gpytorch.mlls import ExactMarginalLogLikelihood
from torch import Tensor

train_X = torch.rand((10, 2))
train_Y = torch.sum(train_X, dim=-1, keepdim=True) + torch.randn(10, 1) * 0.1

gp = SingleTaskGP(train_X, train_Y, outcome_transform=Standardize(m=1))
mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_model(mll)

bounds = Tensor([[0], [1]]).repeat(1, 2)
acqf = qKnowledgeGradient(gp)
fixed_features = {0: 0.5}

solution, value = optimize_acqf(acqf, bounds=bounds, q=1, num_restarts=10, raw_samples=100, fixed_features=fixed_features)

print("solution = %s" % solution)
```

## System Info
Please provide information about your setup, including
- BoTorch Version: up to date with master
- GPyTorch Version: up to date with master
- PyTorch Version: 1.3.1
- Computer OS: macOS Catalina 10.15.2
- Python: 3.7.5 on Anaconda

**Comment by Balandat** [comment]
Thanks for flagging this issue. You're correct, the `gen_batch_initial_conditions` helper that is used internally in `optimize_acqf` doesn't account for the fixed features. The proper thing to do here would be to pass the `fixed_features` dict down to that helper, generate a low-discrepancy set of sample points in `d - n_fix`-dimensional space, where `d` is the dim of the feature space and `n_fix` is the number of fixed features, and then insert the fixed columns at the right place. 

So far this hasn't been high up on the wishlist (not quite sure how important it is), but it should be very easy to implement. We probably won't get to this until after the break, but PRs are always welcome :)

**Comment by saitcakmak** [comment]
@Balandat, I was looking into this and figured that the issue is not specific to `fixed_features` but also applies to `inequality_constraints` and `equality_constraints` as well. It is very straightforward to fix `fixed_features` as you mentioned above. However, things get complicated when you consider all three together. As far as I can tell, there's no readily available function for generating constrained random variables. Any ideas how to generate random / sobol samples subject to multiple `inequality_constraints` and multiple `equality_constraints`? Unless there's a more general use case for such a constrained sampler, this might be more trouble than it's worth.

**Comment by Balandat** [comment]
Thanks @saitcakmak for flagging this. There are two aspects to this problem:
(i) generate random variable that satisfy all constraints used in the initial exploration points (for evaluating the true function)
(ii) generate initial conditions for optimizing the acquisition function on the surrogate model

BoTorch doesn't really have any canned functionality for (i). For that you can do something manually or use Ax (preferred).

A general way to do this is rejection sampling ([here](https://github.com/facebook/Ax/blob/master/ax/models/random/base.py#L108-L119) is where Ax does this), but that may be very sample inefficient if the feasible volume is small relative to the bounding box. Since the feasible set is a polytope, we can instead quite easily use a hit-and-run sampling strategy to generate uniform samples. We have an implementation of this, but have not had the time to put this up as a PR - let me create an issue for this. However, generating quasi-random low-discrepancy samples from a general convex polytope is, as far as I can tell, an open problem.

For (ii) having samples that don't satisfy the constraints is generally not a huge issue, since typically the scipy optimizer can deal with that (usually just projects it onto the feasible set internally). But I do agree that having better initialization would likely improve optimization (e.g. if the feasible set has small volume, the initial conditions projected in the optimizer may all end up on the boundary, which won't be great).

**Comment by saitcakmak** [comment]
I have the same two issues with rejection sampling and letting scipy handle the constraints. I looked into the Ax code you linked, and it seems to only handle inequality constraints. To handle the equality constraints, I was thinking of potentially generating samples at a low dimensional space (of dimension dim - rank(eq_constraints)), and solving for the remaining dimensions; though, I am not sure how efficient this would be.
I am not familiar with hit-and-run sampling. I will read a bit and think about how it could be used here. 

PS, this issue does not relate to a problem I am currently facing. I am trying to find an efficient implementation that could be integrated into BoTorch.

**Comment by Balandat** [comment]
> I was thinking of potentially generating samples at a low dimensional space (of dimension dim - rank(eq_constraints)), and solving for the remaining dimensions;

Yeah that is what you want to do in this case. @jelena-markovic has an implementation of this, we'll hope to clean it up and put it up as a PR soon.

**Comment by saitcakmak** [comment]
Closing this, thanks to #839 (and other related stuff that recently got merged in).
